feat: generate++

mistydemeo opened the pull request:
Closes #371.

This version covers steps 1 and 2, but not 3 and 4 yet.

Gankra commented on cargo-dist/src/errors.rs:
```diff
@@ -159,6 +159,13 @@ pub enum DistError {
         details: semver::Error,
     },
+
+    /// Not an error; indicates that a file's contents differ via --check
```
wdym not an error?

Gankra commented on cargo-dist/src/cli.rs:
```diff
@@ -233,8 +236,33 @@ pub struct InitArgs {
+    pub check: bool,
+}
+
+#[derive(Args, Clone, Debug)]
+pub struct GenerateCiArgs {
+    /// Check if the generated output differs from on-disk config without writing it
+    #[clap(long)]
+    #[clap(default_value_t = false)]
+    pub check: bool,
+}
```
wait... OH this is your preserving the alias (no idea if there's a less painful way to do it, so respect)

mistydemeo replied:
Haha, yeah, that's right.

Gankra commented on cargo-dist/src/main.rs:
```diff
@@ -220,7 +221,25 @@ fn cmd_init(cli: &Cli, args: &InitArgs) -> Result<(), miette::Report> {
 
-fn cmd_generate_ci(cli: &Cli, _args: &GenerateCiArgs) -> Result<(), miette::Report> {
+fn cmd_generate(cli: &Cli, args: &GenerateArgs) -> Result<(), miette::Report> {
+    let modes = if args.mode.is_empty() {
+        vec![GenerateMode::Ci]
+    } else {
+        args.mode.clone()
+    };
```
I'd prefer that this be resolved deeper in the code. `generate`'s library impl should be reading the full config (probably just building a DistGraph) and determining which generate backends are in the "universe" (e.g. "they have ci=github, so I want to generate github ci", "they have installer=msi, so i want to generate wix files", ...).

If the input is an empty array, that's basically saying "no filters, generate everything". If the input isn't empty, that says "do the intersection of everything and this list".

Gankra commented on cargo-dist/src/main.rs:
```diff
@@ -230,7 +249,7 @@ fn cmd_generate_ci(cli: &Cli, _args: &GenerateCiArgs) -> Result<(), miette::Repo
         announcement_tag: cli.tag.clone(),
     };
-    let args = cargo_dist::GenerateCiArgs {};
+    let args = cargo_dist::GenerateCiArgs { check: args.check };
     do_generate_ci(&config, &args)
 }
```
Similarly it would be nice if this was calling `do_generate(&[GenerateStyle::Ci], args.check)` so that the alias boils away immediately.

Gankra commented on cargo-dist/src/backend/ci/github.rs:
```diff
@@ -34,19 +37,43 @@ struct CiTask {
+    let rendered = generate_github_ci(dist)?;
+    // FIXME: should we catch all errors, or only LocalAssetNotFound?
+    let existing = LocalAsset::load_string(ci_file).unwrap_or("".to_owned());
+    Ok(rendered == existing)
```
Hmm, feels like this probably wants to be just returning `Result<(), ...>` and producing the DistError::CheckFileMismatch itself, so that paths/details can be given?

mistydemeo commented on cargo-dist/src/cli.rs:
```diff
@@ -233,8 +236,33 @@ pub struct InitArgs {
+/// Which style(s) of configuration to generate
+#[derive(ValueEnum, Copy, Clone, Debug)]
+pub enum GenerateMode {
+    /// Generate all configuration types (currently: ci)
+    All,
```
Yeah, that's fair. I'll probably switch to just having it be implicit if nothing's specified.

mistydemeo commented:
Updated to add steps 3 and 4 - `allow-dirty` field added to config, added automatic checking during `dist` and manifest commands, and exercised `allow-dirty` in tests.
